--- a/functions.inc.php
+++ b/functions.inc.php
@@ -1244,7 +1244,14 @@ function db_connect ($ignore_errors = 0)
     global $DEBUG_TEXT;
     if ($ignore_errors != 0) $DEBUG_TEXT = '';
     $error_text = '';
-    $link = 0;
+
+    static $link;
+    if (isset($link) && $link) {
+        if ($ignore_errors) {
+            return array($link, $error_text);
+        }
+        return $link;
+    }    $link = 0;
 
     if ($CONF['database_type'] == "mysql") {
         if (function_exists ("mysql_connect")) {
@@ -1412,14 +1419,11 @@ function db_query ($query, $ignore_error
     global $DEBUG_TEXT;
     $result = "";
     $number_rows = "";
-    static $link;
+    $link = db_connect ();
     $error_text = "";
     if ($ignore_errors) $DEBUG_TEXT = "";
 
-    # mysql and pgsql $link are resources, mysqli $link is an object
-    if (! (is_resource($link) || is_object($link) ) ) $link = db_connect ();
-
-    if ($CONF['database_type'] == "mysql") $result = @mysql_query ($query, $link) 
+    if ($CONF['database_type'] == "mysql") $result = @mysql_query ($query, $link)
         or $error_text = "Invalid query: " . mysql_error($link);
     if ($CONF['database_type'] == "mysqli") $result = @mysqli_query ($link, $query) 
         or $error_text = "Invalid query: " . mysqli_error($link);
