Origin: https://github.com/froonix/postfixadmin/commit/ce65c482383554c5f42b81a86132933b433bb67e.patch
Description: https://github.com/postfixadmin/postfixadmin/pull/242
Subject: Fix missing SSL at persistent MySQLi reconnect
Author: Christian Schrötter <cs@fnx.li>
Last-Update: 2019-02-08
Forwarded: not-needed

--- a/functions.inc.php
+++ b/functions.inc.php
@@ -1462,7 +1462,7 @@ function db_connect($ignore_errors = false) {
             if (function_exists("mysqli_real_connect")) {
                 $link = mysqli_init();
                 $link->ssl_set($CONF['database_ssl_key'], $CONF['database_ssl_cert'], $CONF['database_ssl_ca'], $CONF['database_ssl_ca_path'], $CONF['database_ssl_cipher']);
-                $connected = mysqli_real_connect($link, $CONF['database_host'], $CONF['database_user'], $CONF['database_password'], $CONF['database_name'], $CONF['database_port']);
+                $connected = mysqli_real_connect($link, $CONF['database_host'], $CONF['database_user'], $CONF['database_password'], $CONF['database_name'], $CONF['database_port'], null, constant('MYSQLI_CLIENT_SSL'));
                 $is_connected = $connected;
             } else {
                 $error_text .= "<p />DEBUG INFORMATION:<br />MySQLi 5 functions not available! (php5-mysqli installed?)<br />database_type = 'mysqli' in config.inc.php, are you using a different database? $DEBUG_TEXT";
